### PR TITLE
[CDAP-9025] Instead of admin users support admin groups, so that users can be add…

### PR DESCRIPTION
**Pull request by Nishith**
…ed to the admin group to gain privileges and removed from the group if privilege revocation is required.

### CODE REVIEW

File: cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java [hunk 1]
```diff
@@ -900,7 +900,7 @@
       public static final String CACHE_TTL_SECS = "security.authorization.cache.ttl.secs";
       /** Maximum number of entries the authorization cache will hold */
       public static final String CACHE_MAX_ENTRIES = "security.authorization.cache.max.entries";
-      public static final String ADMIN_USERS = "security.authorization.admin.users";
```
Comment (sreevatsanraman) [inline cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java:912]: Are we deprecating this property? Will this break backward compatibility?

Comment (Nishith) [reply cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java:912]: Discussed offline and decided to drop off this property as this is a security hole.

**Comment by Nishith** [comment]
Build: https://builds.cask.co/browse/CDAP-RUT984-3

File: cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java (hunk 1 above)
Comment (rohitsinha54) [reply cdap-common/src/main/java/co/cask/cdap/common/conf/Constants.java:912]: But removing it from here does not resolve security loophole or backward compatibility correct ? We can expect users to specify new groups who will have admin but what about admin.users who have already been granted access. Do we need to drop privileges for them ?

File: cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizationBootstrapper.java [hunk 2]
```diff
@@ -60,9 +60,9 @@
       throw Throwables.propagate(e);
     }
     this.systemUser = new Principal(currentUser, Principal.PrincipalType.USER);
-    this.adminUsers = getAdminUsers(cConf);
-    if (enabled && adminUsers.isEmpty()) {
-      LOG.info("Admin users specified by {} is empty.", Constants.Security.Authorization.ADMIN_USERS);
+    this.adminGroups = getAdminGroups(cConf);
+    if (enabled && adminGroups.isEmpty()) {
+      LOG.info("Admin users specified by {} is empty.", Constants.Security.Authorization.ADMIN_GROUPS);
```
Comment (rohitsinha54) [inline cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizationBootstrapper.java:65]: groups

File: cdap-security-spi/src/main/java/co/cask/cdap/security/spi/authentication/SecurityRequestContext.java [hunk 3]
```diff
@@ -64,9 +65,26 @@ public static void setUserIP(String userIPParam) {
   }
 
   /**
+   * @return the principal type set on the current thread or PrincipalType.USER if not set
```
Comment (rohitsinha54) [inline cdap-security-spi/src/main/java/co/cask/cdap/security/spi/authentication/SecurityRequestContext.java:68]: Why do we need this ? Doesn't the SecurityRequestContext represent the logged in user ? Type is  always user no ?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/AuthorizationBootstrapperTest.java [hunk 4]
```diff
@@ -197,7 +197,8 @@ public Boolean call() throws Exception {
     Assert.assertNotNull(systemDataset);
     // as part of bootstrapping, admin users were also granted admin privileges on the CDAP instance, so they can
     // create namespaces
-    SecurityRequestContext.setUserId(ADMIN_USER.getName());
+    SecurityRequestContext.setUserId(ADMIN_GROUP.getName());
```
Comment (rohitsinha54) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/AuthorizationBootstrapperTest.java:200]: Why set group to userId ? I think you need a mock UserGroupMappingService for the unit test

**Review by rohitsinha54 — CHANGES_REQUESTED** [review]
Some comments. Most importantly we need to make sure that we handle the upgrade properly. if admin.users was configured to user1,user2 and now if we drop this setting and introduce group one we need to handle the privilege which got granted to user1,user2.

**Comment by sreevatsanraman** [comment]
Closing the PR. We will address in 4.3

**Comment by rohitsinha54** [comment]
Note: Please don't delete this remote branch.

**Pull request closed without merging**